File: frontend/src/lib/tdee.ts
// TDEE preview math — Phase 2 Plan 02 (P2-B.2).
//
// EXACT mirror of `backend/app/lib/tdee.py`. The displayed kcal target on
// /dashboard always comes from the server response; this module is only used
// for the live preview during onboarding screen 2 and the /profile edit form.
//
// If FE preview diverges from BE persisted value, that is a sync bug — the
// P2-F.1 end-to-end smoke explicitly compares the two numbers.
//
// Constants MUST stay byte-identical with the Python file:
//   ACTIVITY_FACTORS, KCAL_FLOORS, PROTEIN_FACTOR_G_PER_KG.

export type Sex = "male" | "female";
export type ActivityLevel =
  | "sedentary"
  | "lightly_active"
  | "moderately_active"
  | "very_active"
  | "extra_active";
export type PrimaryGoal = "weight_loss" | "muscle_gain";

export const ACTIVITY_FACTORS: Record<ActivityLevel, number> = {
  sedentary: 1.2,
  lightly_active: 1.375,
  moderately_active: 1.55,
  very_active: 1.725,
  extra_active: 1.9,
};

export const KCAL_FLOORS: Record<Sex, number> = {
  male: 1500,
  female: 1200,
};

export const PROTEIN_FACTOR_G_PER_KG = 1.6;

// Half-up rounding to mirror Python's Decimal(...).quantize(ROUND_HALF_UP).
// JS's Math.round rounds half-to-even for ties at .5 in some implementations
// (engine-dependent), so we implement explicit half-up.
function roundHalfUp(value: number): number {
  return Math.floor(value + 0.5);
}

export function bmrMifflinStJeor(
  sex: Sex,
  weightKg: number,
  heightCm: number,
  age: number,
): number {
  if (sex === "male") {
    return 10 * weightKg + 6.25 * heightCm - 5 * age + 5;
  }
  if (sex === "female") {
    return 10 * weightKg + 6.25 * heightCm - 5 * age - 161;
  }
  throw new Error(`invalid sex: ${sex}`);
}

export function tdee(bmr: number, activityLevel: ActivityLevel): number {
  const factor = ACTIVITY_FACTORS[activityLevel];
  if (factor === undefined) {
    throw new Error(`invalid activity_level: ${activityLevel}`);
  }
  return bmr * factor;
}

export interface KcalTargetResult {
  kcal_target: number;
  floor_hit: boolean;
}

export function dailyKcalTarget(
  sex: Sex,
  weightKg: number,
  heightCm: number,
  age: number,
  activityLevel: ActivityLevel,
  primaryGoal: PrimaryGoal,
): KcalTargetResult {
  if (sex !== "male" && sex !== "female") {
    throw new Error(`invalid sex: ${sex}`);
  }
  if (primaryGoal !== "weight_loss" && primaryGoal !== "muscle_gain") {
    throw new Error(`invalid primary_goal: ${primaryGoal}`);
  }

  const bmr = bmrMifflinStJeor(sex, weightKg, heightCm, age);
  const tdeeKcal = tdee(bmr, activityLevel);
  const target =
    primaryGoal === "weight_loss" ? tdeeKcal - 500 : tdeeKcal + 250;

  const floor = KCAL_FLOORS[sex];
  if (target < floor) {
    return { kcal_target: floor, floor_hit: true };
  }
  return { kcal_target: roundHalfUp(target), floor_hit: false };
}

export function dailyProteinGTarget(
  weightKg: number,
  primaryGoal: PrimaryGoal,
): number | null {
  if (primaryGoal !== "weight_loss" && primaryGoal !== "muscle_gain") {
    throw new Error(`invalid primary_goal: ${primaryGoal}`);
  }
  if (primaryGoal !== "muscle_gain") {
    return null;
  }
  return roundHalfUp(weightKg * PROTEIN_FACTOR_G_PER_KG);
}

File: backend/app/lib/tdee.py
"""TDEE / kcal-target / protein-target pure functions for FitGH.

Phase 2 Plan 02 — Task P2-A.1.

This module is the contract that `backend/app/routes/profile.py` and
`backend/app/routes/weights.py` consume. The frontend has a TS mirror at
`frontend/src/lib/tdee.ts` (D-SHARED-SCHEMA-MANUAL-MIRROR) — the formulas,
activity factors, kcal floors, and protein factor MUST stay identical.
The end-to-end smoke (P2-F.1) compares the FE preview number against the
BE-persisted number; any divergence is a bug.

Formulas:
  - BMR: Mifflin-St Jeor (1990) — NOT Harris-Benedict, NOT Katch-McArdle.
      male:   10*kg + 6.25*cm - 5*age + 5
      female: 10*kg + 6.25*cm - 5*age - 161
  - TDEE: BMR * activity factor (Mifflin's published 1.2..1.9 range).
  - Daily kcal target:
      weight_loss: TDEE - 500  (≈0.5 kg/wk loss, sustainable)
      muscle_gain: TDEE + 250  (≈0.25 kg/wk gain)
      Floor: 1500 male, 1200 female; clamp + set floor_hit=True.
      Round half-up to integer.
  - Daily protein target (muscle_gain only): round(weight_kg * 1.6).
      weight_loss users get None (no protein card on dashboard).
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import Literal, TypedDict

Sex = Literal["male", "female"]
ActivityLevel = Literal[
    "sedentary",
    "lightly_active",
    "moderately_active",
    "very_active",
    "extra_active",
]
PrimaryGoal = Literal["weight_loss", "muscle_gain"]

ACTIVITY_FACTORS: dict[str, float] = {
    "sedentary": 1.2,
    "lightly_active": 1.375,
    "moderately_active": 1.55,
    "very_active": 1.725,
    "extra_active": 1.9,
}

KCAL_FLOORS: dict[str, int] = {
    "male": 1500,
    "female": 1200,
}

PROTEIN_FACTOR_G_PER_KG: float = 1.6


class KcalTargetResult(TypedDict):
    kcal_target: int
    floor_hit: bool


def _round_half_up(value: float) -> int:
    """Round half-up to integer; Python's default banker's rounding is wrong here."""
    return int(Decimal(str(value)).quantize(Decimal("1"), rounding=ROUND_HALF_UP))


def bmr_mifflin_st_jeor(
    sex: Sex,
    weight_kg: float,
    height_cm: float,
    age: int,
) -> float:
    """Mifflin-St Jeor basal metabolic rate (kcal/day) — returns float for chained math."""
    if sex == "male":
        return 10 * weight_kg + 6.25 * height_cm - 5 * age + 5
    if sex == "female":
        return 10 * weight_kg + 6.25 * height_cm - 5 * age - 161
    raise ValueError(f"invalid sex: {sex!r}")


def tdee(bmr: float, activity_level: ActivityLevel) -> float:
    """Total Daily Energy Expenditure = BMR * activity factor."""
    if activity_level not in ACTIVITY_FACTORS:
        raise ValueError(f"invalid activity_level: {activity_level!r}")
    return bmr * ACTIVITY_FACTORS[activity_level]


def daily_kcal_target(
    sex: Sex,
    weight_kg: float,
    height_cm: float,
    age: int,
    activity_level: ActivityLevel,
    primary_goal: PrimaryGoal,
) -> KcalTargetResult:
    """Compute the user's daily kcal target with floor enforcement."""
    if sex not in ("male", "female"):
        raise ValueError(f"invalid sex: {sex!r}")
    if primary_goal not in ("weight_loss", "muscle_gain"):
        raise ValueError(f"invalid primary_goal: {primary_goal!r}")
    # tdee() validates activity_level

    bmr = bmr_mifflin_st_jeor(sex, weight_kg, height_cm, age)
    tdee_kcal = tdee(bmr, activity_level)

    if primary_goal == "weight_loss":
        target = tdee_kcal - 500
    else:  # muscle_gain
        target = tdee_kcal + 250

    floor = KCAL_FLOORS[sex]
    floor_hit = target < floor
    if floor_hit:
        return {"kcal_target": floor, "floor_hit": True}
    return {"kcal_target": _round_half_up(target), "floor_hit": False}


def daily_protein_g_target(
    weight_kg: float,
    primary_goal: PrimaryGoal,
) -> int | None:
    """Protein target in grams/day. Only meaningful for muscle_gain users."""
    if primary_goal not in ("weight_loss", "muscle_gain"):
        raise ValueError(f"invalid primary_goal: {primary_goal!r}")
    if primary_goal != "muscle_gain":
        return None
    return _round_half_up(weight_kg * PROTEIN_FACTOR_G_PER_KG)
